generators/E906LegacyVtxGen/SQPrimaryVertexGen.C

```diff
@@ -0,0 +1,270 @@
+  beamProfile->SetParameter(1, rc->get_DoubleFlag("SIGX_BEAM"));
+  beamProfile->SetParameter(2, rc->get_DoubleFlag("Y_BEAM"));
+  beamProfile->SetParameter(3, rc->get_DoubleFlag("SIGY_BEAM"));
+
+  rndm.SetSeed(PHRandomSeed());
+
+  //If inited from a root file, then skip this step, so that it can be used independent of the Fun4All
+  if(geoManager == nullptr) geoManager = PHGeomUtility::GetTGeoManager(topNode);
+
+  //Read the default 
+  defaultMatProf = new MaterialProfile;
+  fillMaterialProfile(defaultMatProf, 0., 0.);
+
+  inited = true;
+}
+
+void SQPrimaryVertexGen::fillMaterialProfile(MaterialProfile* prof, double xvtx, double yvtx)
+{
+  //std::cout << xvtx << "  " << yvtx << std::endl;
+  //retrieve the z position of all geometry boundaries
+  recoConsts* rc = recoConsts::instance();
+
+  double z[100];
+  int nBoundaries = 0;
+
+  double eps = 1.E-6;
+  double z_curr = -800.;//not beyond collimator(z= -602.36 cm,lenth=121.92 cm) @Abi //-500.;
+  z[0] = z_curr;
+  while(z_curr < 503.)
+    {
+      geoManager->IsSameLocation(xvtx, yvtx, z_curr, true);
+      geoManager->SetCurrentDirection(0., 0., 1.);
+
+      TGeoNode* node = geoManager->FindNextBoundary(600.);
+      double z_step  = geoManager->GetStep();
+
+      if(z_step > eps)
+	{
+	  z_curr += z_step;
+	  z[++nBoundaries] = z_curr; 
+	}
+      else
+	{
+	  z_curr += eps;
+	}
+    }
+
+  //use the z boundaries to create interactibles
+  for(int i = 0; i < nBoundaries; ++i)
+    {
+      double z_center = 0.5*(z[i] + z[i+1]);
+      if(z_center > 503.) continue;
+ 
+      SQBeamlineObject obj(geoManager->FindNode(xvtx, yvtx, z_center)->GetVolume()->GetMaterial());
+      TString name = geoManager->FindNode(xvtx, yvtx, z_center)->GetVolume()->GetName();
+      obj.z_up = z[i];
+      obj.z_down = z[i+1];
+      obj.z0 = z_center;
+      obj.length = z[i+1] - z[i];
+    
+      //TString name = obj.name;    
+      //std::cout<<" Name of beam volume "<<name<<std::endl;  
+
+      if (_targetOnlyMode){
+	if(! name.Contains("Target")) continue;
+      }
+
+      if (_dumpOnlyMode){
+	if(! name.Contains("fmag")) continue;
+      }
+
+      prof->interactables.push_back(obj);
+    }
+
+  //calculate all related properties
+  prof->calcProb();
+
+  // for(int i = 0; i < prof->nPieces; ++i)
+  // {
+  //   std::cout << prof->interactables[i] << std::endl;
+  // }
+  // return 0;
+}
+
+
+TVector3 SQPrimaryVertexGen::generateVertex()
+{
+  if(!inited) init();
+
+  double xvtx, yvtx, zvtx;
+  generateVtxPerp(xvtx, yvtx);
+
+  bool inTarget = ((xvtx-targetX0)*(xvtx-targetX0)/targetSX/targetSX + (yvtx-targetY0)*(yvtx-targetX0)/targetSY/targetSY) < 1.;
+  MaterialProfile* activeProf;
+  if(!inTarget)
+    {
+      
+      activeProf = new MaterialProfile;
+      fillMaterialProfile(activeProf, xvtx, yvtx);
+    }
+  else
+    {
+      activeProf = defaultMatProf;
+    }
+
+  int index = activeProf->findInteractingPiece(rndm.Rndm());
```

Comment:
This line (or next) gets into trouble when the number of interacting pieces is zero.  Here we should repeat the generation of xvtx and yvtx until the number becomes non-zero.

Reply:
Thank you Kenichi for the suggestion. I have deleted the portion inside the "funcBeamProfile" and added condition for generating the xvtx and yvtx inside the target area for _targetOnlyMode. I have pushed the commits with your suggestions.

Reply:
I would like to take a more-general method for this.  I understand that the target-only mode and other special modes are reflected in `MaterialProfile` by `fillMaterialProfile()`.  Thus `generateVertex()` need not worry about the mode (`_targetOnlyMode`) but generates the x,y vertex until `activeProf->interactables.size() > 0` in principle.

I'm afraid that the present method is so specific to cause a confusion/bug in future.  

Can you also make `_targetOnlyMode` and `_dumpOnlyMode` not static?

Reply:
I have changed `_targetOnlyMode` and `_dumpOnlyMode` to non-static one (I forgot why I set them like that in first place). Thank you for your pointing me that out.

Also, I will try to look into more-general method for this.

Reply:
Current implementation of target only and dump only mode is not very good since it introduces expensive string comparison many times during the generation process. I would use start/stop z position for the geometry navigator instead to limit the search for the interacables to target/dump only locations. 

Let me push a change to this PR today to address this issue.